kie-server-parent/kie-server-services/kie-server-services-openshift/src/main/java/org/kie/server/services/impl/OpenShiftStartupStrategy.java

```diff
@@ -0,0 +1,203 @@
+import io.fabric8.openshift.client.OpenShiftClient;
+import org.kie.server.api.KieServerConstants;
+import org.kie.server.api.model.KieContainerResource;
+import org.kie.server.controller.api.model.KieServerSetup;
+import org.kie.server.services.api.StartupStrategy;
+import org.kie.server.services.impl.storage.KieServerState;
+import org.kie.server.services.impl.storage.cloud.CloudClientFactory;
+import org.kie.server.services.impl.storage.cloud.KieServerStateCloudRepository;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OpenShiftStartupStrategy implements StartupStrategy {
+
+    private static final String KIE_SERVER_ROLLOUT_IN_PROGRESS = "kieserver-rollout-in-progress";
+    private static final String KIE_SERVER_INSTANCE_ID = "kieserver-" + UUID.randomUUID().toString();
+    private static final String ROLLOUT_TRIGGER_TIMESTAMP = "services.server.kie.org/openshift-startup-strategy.changeTimestamp";
+    private static final Logger logger = LoggerFactory.getLogger(OpenShiftStartupStrategy.class);
+
+    private static Supplier<OpenShiftClient> clouldClientHelper = () -> 
+        (OpenShiftClient) (new CloudClientFactory() {}).createAdaptiveClient();
+
+    private static class WatchRunner implements Runnable {
+
+        private boolean isWatchRunning = true;
+        private String kieServerId;
+
+        public WatchRunner(String serverId) {
+            kieServerId = serverId;
+        }
+
+        @Override
+        public void run() {
+            try (OpenShiftClient client = clouldClientHelper.get()) {
+
+                logger.info("Watching ConfigMap in namespace {}", client.getNamespace());
+
+                try (Watch watchable = client.configMaps().watch(new Watcher<ConfigMap>() {
+
+                    @Override
+                    public void eventReceived(Action action, ConfigMap kieServerState) {
+
+                        logger.info("Event - Action: {}, {} on ConfigMap ",
+                                    action, kieServerState.getMetadata().getName());
+
+                        DeploymentConfig dc = client.deploymentConfigs().withName(kieServerId).get();
+
+                        if (kieServerId.equals(kieServerState.getMetadata().getName()) 
+                                && action.equals(Action.MODIFIED) 
+                                && "True".equals(dc.getStatus().getConditions().get(0).getStatus()) 
+                                && triggerRollout(client, kieServerId)) {
+
+                            ObjectMeta md = dc.getSpec().getTemplate().getMetadata();
+                            Map<String, String> ann = md.getAnnotations() == null ? new HashMap<>() : md.getAnnotations();
+                            md.setAnnotations(ann);
+                            ann.put(ROLLOUT_TRIGGER_TIMESTAMP,
+                                    ZonedDateTime.now().format(DateTimeFormatter.ISO_INSTANT));
+                            client.deploymentConfigs().createOrReplace(dc);
```

Comment:
Yes. The lifecycle of the KieServer Pods is controlled by the associated DeploymentConfig (DC) which has configured with automatic rollout triggers. Once OpenShift detects any change made to that DC, (In our case, it is by touching/adding an annotation) it will terminate and recreate Pods according to the specified strategy (Recreate or Rolling) 

https://docs.openshift.com/container-platform/3.9/dev_guide/deployments/deployment_strategies.html#rolling-strategy

Rolling strategy will provide a non-interrupt experiences to use and service consumer.

Reply:
ok, thanks